Update RISCV-64 sleigh files to support vector, bit manipulation, and crypto extensions

thixotropist opened the pull request:
Add several RISCV Instruction Set extensions to Ghidra, following discussion https://github.com/NationalSecurityAgency/ghidra/discussions/5744. This pull request tracks the tip of the binutils testsuite for vector, bitmap, and crypto instructions. You can verify the content by importing sample binaries from https://github.com/thixotropist/ghidra_import_tests. Import the RISCV-64 gas test suite, assemble to binary, then iterate on the Ghidra sleigh files until Ghidra and objdump give essentially the same disassembled output.

The sleigh files do not yet include pcode semantics. Recent updates to GCC-14 and libssl using RISCV vector and crypto extensions may give us sample binaries to work with, to see what pcode semantics actually add value with complex instructions like these.

jobermayr commented:
To fix build errors:
```
diff --git a/Ghidra/Processors/RISCV/certification.manifest b/Ghidra/Processors/RISCV/certification.manifest
index 569138783..b498068db 100644
--- a/Ghidra/Processors/RISCV/certification.manifest
+++ b/Ghidra/Processors/RISCV/certification.manifest
@@ -40,6 +40,9 @@ data/languages/riscv.rvc.sinc||GHIDRA||||END|
 data/languages/riscv.rvv.sinc||GHIDRA||||END|
 data/languages/riscv.table.sinc||GHIDRA||||END|
 data/languages/riscv.zi.sinc||GHIDRA||||END|
+data/languages/riscv.zvbb.sinc||GHIDRA||||END|
+data/languages/riscv.zvkng.sinc||GHIDRA||||END|
+data/languages/riscv.zvksg.sinc||GHIDRA||||END|
 data/languages/riscv32-fp.cspec||GHIDRA||||END|
 data/languages/riscv32.cspec||GHIDRA||||END|
 data/languages/riscv32.dwarf||GHIDRA||||END|
```

thixotropist commented:
I expect to fill in some gaps in this PR shortly.  Scalar crypto extensions were skipped even though vector crypto extensions were added.  openssl can use RISCV scalar crypto AES extension instructions but not (yet?) the vector crypto extensions.  I also hope to add minimalist pcode semantics to allow decompilation of the *simplest* GCC-14 RISCV builtin intrinsic vector function examples - as used in  `rvv_memcpy`, `rvv_strncpy`, `rvv_matmul`, and `rvv_reduce`. 

Ghidra developers will have some serious design questions to thrash out when GCC-14 autovectorization support lands some time next year.

mumbel commented on Ghidra/Processors/RISCV/data/languages/riscv.instr.sinc:
```diff
@@ -46,9 +46,23 @@
+
+# Thead semi's extensions currently recognized by binutils objdump
+#   and documented in https://github.com/T-head-Semi/thead-extension-spec/releases/download/2.0.0/xthead-2022-09-05-2.0.0.pdf
+@include "riscv.xthead.sinc"
```
This should be guarded by `ifdef`, with the `define` in a new xhead slaspec file

thixotropist replied:
That's reasonable.  Does the new xthead slaspec file get named in `riscv.ldefs` so the user can invoke it, or do you suggest we generalize `riscv.opinion` to look for the `Tag_RISCV_arch` ELF attribute, recognize the current composition of extensions,  and set finer-grain inclusion tags? 
```
Tag_RISCV_arch: "rv64i2p1_m2p0_a2p1_f2p2_d2p2_v1p0_zicsr2p0_zifencei2p0_zmmul1p0_zve32f1p0_zve32x1p0_zve64d1p0_zve64f1p0_zve64x1p0_zvl128b1p0_zvl32b1p0_zvl64b1p0_xtheadfmemidx1p0"
```

Apparently binutils 2.41 concatenates all of the `march` extensions passed to gcc, so the tail end of this attribute reads something like:

"This binary requires an ISA supporting:"

* `x` - a vendor specific Instruction Set Architecture module not currently part of a proposed standard profile
* `thead` - the lower case vendor name publishing the extension set
* `fmemidx` - the extension set module name
* `1p0` - the version of this extension set, likely `1.0`

Composing standard and vendor extension profiles raises lots of Ghidra import design questions.  Apparently questions the binutils team has already addressed - do their answers work for Ghidra?

thixotropist commented:
THead extensions are now collected into separate slaspec files, which are now referenced in `riscv.ldefs`.  I've only tested the 64 bit version.  Thanks to @mumbel for making this suggestion.

* each of the 10  THead extensions is guarded by its own versioned ifdef flag derived from the name embedded in `Tag_RISCV_arch`.  That should give a bit more granular control.
* ` riscv.opinion` remains unchanged. The presence of ISA extensions is identified in ELF by appending to `Tag_RISCV_arch`, rather than setting a bit in a fixed ELF header bitfield `e_flags`.
* users will only see the THead language option on import or change-language if they show all RISCV languages, not just the recommended language based on e_machine and e_flags

madushan1000 commented:
@thixotropist I stubbed out some risc-v packed simd instructions and implemented some thead instructions [here](https://github.com/madushan1000/ghidra/commits/isa_ext) to fix some issues had with my work, feel free to cherry-pick the commits if you want.

thixotropist commented:
@madushan1000: Those look good - I'll be happy to cherrypick them into the branch.  Have you any suggestions for RISCV integration tests to add to https://github.com/thixotropist/ghidra_import_tests?  It's currently very weak in 32 bit and microcontroller exemplars, as I've been leaning towards linux-capable 64 bit examples.

madushan1000 commented:
This sdk I'm working with has a bunch of rv32 examples, https://github.com/bouffalolab/bouffalo_sdk/tree/master/examples.
bl602, bl70x have sifive e24 cores. and bl61x and bl808 has various t-head cores. the examples have a small readme with the build instructions.

jobermayr commented on Ghidra/Processors/RISCV/data/languages/riscv.rvv.sinc:
```diff
@@ -3393,7 +3394,7 @@ define pcodeop vsetvli_e64m1tumu;
 :vsetvli   rd, rs1, "e32,mf2,tu,ma"   is op2030=0x097 & op3131=0x0 & rs1 & op1214=0x7 & rd & op0006=0x57 {rd=vsetvli_e32mf2tuma(rs1);}
 :vsetvli   rd, rs1, "e64,m1,tu,ma"    is op2030=0x098 & op3131=0x0 & rs1 & op1214=0x7 & rd & op0006=0x57 {rd=vsetvli_e64m1tuma(rs1);}
-:vsetvli   rd, rs1, "e32,mf8,tu,ma"   is op2030=0x09d & op3131=0x0 & rs1 & op1214=0x7 & rd & op0006=0x57 {rd=vsetvli_e64mf8tuma(rs1);}
+:vsetvli   rd, rs1, "e32,mf8,tu,ma"   is op2030=0x09d & op3131=0x0 & rs1 & op1214=0x7 & rd & op0006=0x57 {rd=vsetvli_e32mf8tuma(rs1);}
```
missing "define pcodeop vsetvli_e32mf8tuma" leads to compile errors:
```
> Task :RISCV:sleighCompile
Compiling ./data/languages/riscv.ilp32d.slaspec:
riscv.rvv.sinc:3397: unknown macro, userop, or specific symbol 'vsetvli_e32mf8tuma' in macro, user operation, or subpiece application
```

thixotropist replied:
Thanks!  This should now be fixed with ffb5411.

Any comments on other vset* patterns that might be found in common exemplars would be appreciated.  I'm grinding through autovectorized vector instruction sequences generated from https://github.com/ggerganov/llama.cpp currently, and will likely try and make the sinc pcodeop names align a bit better with the gcc riscv intrinsics defined in https://github.com/riscv-non-isa/rvv-intrinsic-doc.

jobermayr commented:
```
diff --git a/Ghidra/Processors/RISCV/data/languages/riscv.zfh.sinc b/Ghidra/Processors/RISCV/data/languages/riscv.zfh.sinc
index 696766a208..f8e0c69a4b 100644
--- a/Ghidra/Processors/RISCV/data/languages/riscv.zfh.sinc
+++ b/Ghidra/Processors/RISCV/data/languages/riscv.zfh.sinc
@@ -10,7 +10,7 @@
 :flh frd,immI(rs1) is immI & frd & rs1 & op0001=0x3 & op0204=0x1 & op0506=0x0 & funct3=0x1
 {
     local ea:$(XLEN) = immI + rs1;
-    fassignS(frd, *[ram]:2 ea);
+    fassignS(frd, *[ram]:4 ea);
 }
 
 # Half-Precision Store Instruction
```
to fix
```
> Task :RISCV:sleighCompile FAILED

Compiling ./data/languages/riscv.lp32qingke.slaspec:
5 NOP constructors found
Use -n switch to list each individually
riscv.table.sinc:103: Size restriction error in table 'instruction' in constructor at riscv.zfh.sinc:10
  Problem with 'frd' in 'Copy(=)' operator
  Input and output sizes must match; {type=real value_real=0x2 spaceid=null} != {type=handle value_real=0x0 spaceid=null}
No output produced
```

thixotropist commented:
@jobermayr: Thanks for the fix. I'm pushing your patch

jobermayr commented on Ghidra/Processors/RISCV/data/languages/riscv.zfh.sinc:
```diff
@@ -0,0 +1,249 @@
+	fassignS(frd, tmp);
+}
+
+:fcvt.w.h rdW,frs1S,FRM is frs1S & FRM & rdW & op0001=0x3 & op0204=0x4 & op0506=0x2 & funct7=0x62 & op2024=0x0
```
https://github.com/NationalSecurityAgency/ghidra/pull/6492/files#diff-f74ddc93319f068d960c0557be5af9198befd17809b4144ba1f50653c199766eL44

jobermayr replied:
Does it make sense to not using rdW Here as well?

thixotropist replied:
Thanks - I'll take a look at this tomorrow.  I want to look at how qemu handles similar riscv instructions.

Have you any thoughts on how the semantics of *vector* floating point conversion instructions should be handled?  Element size is passed as often-changed SEL context variables, not in the instructions themselves.

Qemu source code spends about 24K lines of C to describe RISCV instruction semantics.  I'm not sure how much sleigh pcode that implies, or how to finish what we're starting.

gamelaster commented:
I just tried latest version of [thixotropist:isa_ext](https://github.com/thixotropist/ghidra/tree/isa_ext) on BL808 BootROM, which is E907 (rv32 thead), and so far, everything looks okay. I still need to play with it more, but so far it was enough for everything I needed. Thanks everyone for this effort, I hope it will be possible to merge it at some point.

thixotropist commented:
Thanks for the comments - I'd like to see this merged too. The PR may still be in triage because it implicitly makes a *lot* of design decisions regarding pcode op typing and emulation, as well as ISA extension handling.  The developers may need more discussion - public and internal - before they are willing to go down that path.

The current state of the PR is stable.  There are some newer RISCV ISA extensions for fractional floating point ops and saturating math - I don't plan on adding these to the existing PR, so it can be reviewed as is.

As a discussion example, what does the Ghidra community want to see in the decompiler window when working with functions like:

```c
// compiled with RISCV march=rv64gcv, -O3, and -ffast-math
void test_1_ref(unsigned long long *in, unsigned long long *out, unsigned int size)
{
    int i;
    int upper_index = size - 1;
    for (i=0; i < size; i++) {
        out[i] = in[upper_index - i];
    }
}
```
SIMD or vector extensions can turn simple loops over structures into something not so simple.  Type inference, compilable C extraction, and emulation in general are all abandoned with the design approach used in this PR.

thixotropist commented:
Is there anything we can do to help @GhidorahRex  with this review? For instance
* Documentation
* Test cases
* Design notes and tradeoff analyses

thixotropist commented:
@jobermayr is correct.  I've added guards as suggested, with additional guards sensitive to quad floating point support.  I'll commit these changes shortly.  In general, I've used single-precision registers for half-precision FP ops and double-precision registers for quad-precision FP ops as interim semantics.

This raises more general questions:

* What set of RISCV language definitions and test object data should we be using in testing?  I'm concentrating on RISCV-64 processors with vector 1.0 and RVA-23 support.  I'm happy to test against other platforms if we have a consensus.
    * test cases in the style of `ghidra/Ghidra/Features/Decompiler/src/decompile/datatests/*.xml` would be nice
* What's a better approach to half-precision FP semantics? There are RISCV extensions defined for both IEEE and bfloat types of half-precision numbers, and no current support for either within the Ghidra decompiler or sleigh builtins.

emteere commented:
Would the community ( @mumbel @thixotropist ) have any objection to removing the context register from attachment to the MISA register?

In practice, the context register can be used for configuration of a processor, but tying it to a real register with a restricted size and set bitfields won't allow you to add other types of context that might be necessary for disassembly or other purposes.

The normal reason to put these processor capability setting bits into the context would be if they were going to be used to configure the processor and restrict disassembly to only the instructions available on the particular processor variant.  It looks like maybe this was a considered implementation direction at some point.  The current implementation splits the various extensions into .sinc files and uses includes to configure the processor to restrict disassembly.

thixotropist commented:
@jobermayr - thanks for the heads-up.  I've got a draft rebase building, and will commit it in a few days after I clean up some of the SLEIGH definitions.  I'll also remove the thead extensions from this PR.

gamelaster commented:
I really hope that support of unratified 0.7 vector extension and T-Head ISA will be kept available. Sadly, there are too many designs already done with those extensions.

jobermayr commented:
> @GhidorahRex - I updated my fork of master from the primary Ghidra repo, squash-merged my isa_ext branch, then pushed that back to my master fork. I _think_ that's what you requested...

@thixotropist Not really. What is wanted is on your `isa_ext` branch sth. like:
```
git reset --hard master
pushd ..
wget https://github.com/NationalSecurityAgency/ghidra/pull/5778.patch
wget https://github.com/NationalSecurityAgency/ghidra/pull/5778.diff
popd
patch -p1 --merge <../5778.diff
git add .
git commit -m "$(sed -n '/^Subject:/, /^---$/p' ../5778.patch | sed -e '/^$/d; /^---$/d' -e "s;^Subject: \[PATCH.*\] ;\n;g")"
git push -f
rm ../5778.*
```

thixotropist committed:
Squashed commit of the following:

commit 03363a971df6fe4f174a40e1bbbfc36f1b6c1c91
Merge: 239725a09a 16350d99f4
Author: thixotropist <thixotropist@proton.me>
Date:   Mon Dec 15 15:15:27 2025 -0500

    Merge branch 'master' into isa_ext

commit 239725a09ac4ac801c0b6b03ccd2c53faeeafcbf
Merge: af6f6da8c5 7bed72e389
Author: thixotropist <thixotropist@proton.me>
Date:   Sun Dec 14 14:10:38 2025 -0500

    Merge branch 'master' into isa_ext

commit af6f6da8c59c218cc1905914faa6491c93651c9b
Author: thixotropist <thixotropist@proton.me>
Date:   Sun Dec 14 14:09:38 2025 -0500

    RISC-V vcompress instruction semantics fix

The semantic pcodeop section of many RISC-V instructions needs to
show
    that the output vector register is also a dependent input register.

commit 589ad1a8b9f8e3033f99e6d3f7bc47ffa795c8f6
Author: thixotropist <thixotropist@proton.me>
Date:   Mon Nov 17 08:00:59 2025 -0500

    Restore RISCV T-Head instruction extension support

commit 7e8e7ea18e40be7b314765a337fcc4012c91524e
Merge: 0cbd6af3fb be482754a7
Author: thixotropist <thixotropist@proton.me>
Date:   Thu Nov 13 13:23:54 2025 -0500

    Merge branch 'master' into isa_ext

commit 0cbd6af3fbcd72937a22047aa019dd598b9d626d
Merge: 11aabf63f4 12a8db6195
Author: thixotropist <thixotropist@proton.me>
Date:   Sun Jul 20 09:03:59 2025 -0400

    Merge branch 'master' into isa_ext

commit 11aabf63f4934895fc2b2de6ca9e03aaa0d31a23
Author: thixotropist <thixotropist@proton.me>
Date:   Fri Jul 11 14:00:53 2025 -0400

    add missing input parameters

    Some RISCV vector instructions read the output register.

commit 2b8ab41238fca81a9ae8922c8bbcaae73e9d40af
Merge: 03d51ba85c e63d97b112
Author: thixotropist <thixotropist@proton.me>
Date:   Fri Jul 11 13:46:10 2025 -0400

    Merge branch 'master' into isa_ext

commit 03d51ba85c45042b6dcfcb1776058d754789286e
Merge: cc77553056 7482131bcc
Author: thixotropist <thixotropist@proton.me>
Date:   Tue Jun 24 21:21:01 2025 -0400

    Merge 11.5_DEV branch 'master' into isa_ext

commit cc7755305623d46e7f0c2683d97df05402c30388
Author: thixotropist <thixotropist@proton.me>
Date:   Fri May 9 11:01:52 2025 -0400

    Identify in/out parameters to support dependancy analysis

commit 92968acc01d3505ec39aa33d7271742329d2d773
Merge: dfec2c3c81 eaf6c42a7e
Author: thixotropist <thixotropist@proton.me>
Date:   Thu May 8 09:48:35 2025 -0400

    Merge branch 'master' into isa_ext

commit dfec2c3c81dbb653f1dea9da49c9b0c6ac59037e
Author: thixotropist <thixotropist@proton.me>
Date:   Wed Mar 19 12:08:05 2025 -0400

    Move output parameters out of pcode op's input parameter signature

commit e80ddce52a085e72a8ca8d37fa259e8ca65e0caf
Merge: f0c8cc8cbc ecfd6d39d8
Author: thixotropist <thixotropist@proton.me>
Date:   Wed Mar 19 08:13:12 2025 -0400

    Merge branch 'master' into isa_ext

commit f0c8cc8cbc5f0a3597af33599d67d8b16b670196
Author: thixotropist <thixotropist@proton.me>
Date:   Sat Feb 15 10:20:30 2025 -0500

* Correct guards to allow builds on RISCV platforms without double
precision FP
    * Document RISCV datatests
    * Remove unneeded fp register definition

commit 3aae55417c2ef87f1957f6e794a4cd90fe50c79b
Author: thixotropist <thixotropist@proton.me>
Date:   Wed Feb 12 08:19:50 2025 -0500

    Add initial support for RISCV extension zfa in the disassembly view.
The associated semantics are only suggestive of the action taken by
these instructions

commit fac48f1c60d5701bb38c37c9d7ad2baf6f0b9709
Merge: 3efecd1d5f eaa8aeb0c8
Author: thixotropist <thixotropist@proton.me>
Date:   Wed Feb 12 08:10:23 2025 -0500

    Merge branch 'master' into isa_ext

commit 3efecd1d5f0402e0c6facef69b0c43af22429f9e
Merge: 0d2294a989 cdfcc71255
Author: thixotropist <thixotropist@proton.me>
Date:   Thu Feb 6 07:51:34 2025 -0500

    Merge branch 'master' into isa_ext

commit 0d2294a989065ca8db5230c07d9cc217e3eab78d
Author: thixotropist <thixotropist@proton.me>
Date:   Thu Feb 6 07:42:53 2025 -0500

Add two riscv64 decompiler tests, update isa_ext docs to current
Fedora release

commit 513153c071931758d358ad1c435b04f38b7fe405
Merge: 9a1edb68dd 7a288b3b8c
Author: thixotropist <thixotropist@proton.me>
Date:   Sun Jan 12 07:08:41 2025 -0500

    Merge branch 'master' into isa_ext

commit 9a1edb68dd7ec379ae6c59f06cbf3c188227c886
Merge: 3a2bf919c3 d87b514baa
Author: thixotropist <thixotropist@proton.me>
Date:   Fri Jan 10 10:08:17 2025 -0500

    Merge branch 'master' into isa_ext

commit 3a2bf919c3aa9798c085bc6f182ea17816c56e32
Merge: a59ac6fd22 14527b015e
Author: thixotropist <thixotropist@proton.me>
Date:   Wed Dec 25 12:16:58 2024 -0500

    Merge branch 'master' into isa_ext

commit a59ac6fd22bb3926682b349c467322c11fbb668a
Author: thixotropist <thixotropist@proton.me>
Date:   Sun Dec 15 11:10:56 2024 -0500

    Allow Python 3.13

commit 3b8de7c6cef2992e047155c119c2f001d49aae69
Merge: 06a4981db6 97aadfa05a
Author: thixotropist <thixotropist@proton.me>
Date:   Sun Dec 15 10:43:36 2024 -0500

    Merge branch 'master' into isa_ext

commit 06a4981db6924158852f1c6a0e7cf7d2d8fd5246
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Nov 18 08:54:19 2024 -0500

    allow use of python 3.13

commit faf08bd8cb6fc5b603d27a331b2a08260718f691
Merge: 21bcbf724e 1df4527dd3
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sun Nov 17 12:14:40 2024 -0500

    Merge branch 'master' into isa_ext

commit 21bcbf724e5a498eca0e5a98ba1f66becbb565f5
Merge: 639c0576e5 7e03b30b5b
Author: thixotropist <97179149+thixotropist@users.noreply.github.com>
Date:   Sat Oct 19 10:36:28 2024 -0400

    Merge pull request #1 from cyrozap/fix-th.srX

    Fix store sizes for several instructions

    (My apologies for not noticing the pull request earlier)

commit 639c0576e5e63bc62bf3924d23f580f5be13e7bb
Merge: 48330a6529 990479f79b
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat Oct 19 10:26:27 2024 -0400

    Merge branch 'master' into isa_ext

commit 48330a6529376f89e223765f285c0247be6126cf
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat Oct 19 10:18:24 2024 -0400

    Correct decompilation of zext.h to show the correct register

commit 64ce151e56be5adb2b1974671aec0dfde2c76937
Merge: 2497f18be1 59f7b60a37
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat Sep 28 10:02:57 2024 -0400

    Merge branch 'master' into isa_ext

commit 2497f18be1ac60780e4ca1c40f4088e3157a04e0
Merge: 1241a40ff5 4eefa887fb
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Sep 4 10:31:29 2024 -0400

    Merge branch 'master' into isa_ext

commit 1241a40ff59504b65bfbbe90e989d89f329058b0
Merge: bd8f1ab238 0857d52681
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Jul 18 06:30:51 2024 -0400

    Merge branch 'master' into isa_ext

commit bd8f1ab238a3ce018f33a496886cfde6013dd0d0
Merge: b6c82beb11 dec414e11b
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Jul 10 07:08:02 2024 -0400

    Merge branch 'master' into isa_ext

commit b6c82beb1104763b7ed38d0c423b2cfa11599930
Merge: 459cebe14e 93eed8011c
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sun Jun 30 12:34:18 2024 -0400

    Merge branch 'master' into isa_ext

commit 7e03b30b5bbbbe31cc5fd8f06e39be49dc8506e4
Author: cyrozap <cyrozap@gmail.com>
Date:   Sat Jun 15 18:46:59 2024 -0500

    Fix store sizes for th.fsrw and th.fsurw

commit 1fe84bbadad58941bc86e61e93bc86cd68fc959c
Author: cyrozap <cyrozap@gmail.com>
Date:   Sat Jun 15 18:46:20 2024 -0500

    Fix store sizes for th.swd

commit 244b5a73029d2af32d2c8023700d7c998d65e467
Author: cyrozap <cyrozap@gmail.com>
Date:   Sat Jun 15 18:43:50 2024 -0500

    Fix store sizes for th.surb, th.surh, and th.surw

commit a74dd20452086307780c5cb42f5acc9efb451fc8
Author: cyrozap <cyrozap@gmail.com>
Date:   Sat Jun 15 18:40:26 2024 -0500

Fix store sizes for th.sbia, th.sbib, th.shia, th.shib, th.swia,
th.swib

commit ac0777528a56a81433d0a9198968295419952b96
Author: cyrozap <cyrozap@gmail.com>
Date:   Sat Jun 15 18:02:17 2024 -0500

    Fix store sizes for th.srb, th.srh, and th.srw

commit 459cebe14ef19ea9fd7a76bad15e82f18940035c
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Jun 12 13:33:41 2024 -0400

Allow for 32 bit floating point registers when converting 16 bit
floats

commit bfb6c8b68e845d4dd6eba7bf9c6d47e465e461eb
Merge: ad4b229e35 715a8ba335
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Jun 12 12:11:46 2024 -0400

    Merge branch 'master' into isa_ext

commit ad4b229e35351930cbef1597018153cbbd81334e
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Jun 10 09:39:36 2024 -0400

    cleanup 16 bit floating point riscv support

commit 80a3ecdf01a67f3987623d0ecb21fc156fb3981d
Merge: 22f1a78bdc fb844be1f6
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat May 25 12:16:26 2024 -0400

    Merge branch 'master' into isa_ext

commit 22f1a78bdca93cdb00b6704056a6477059c5ce70
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat May 25 11:56:13 2024 -0400

    Cleanup RISCV half-precision floating point support

commit 22e145b992254835e247f2946dcbee6bda911abb
Merge: 2502672f19 9911db9828
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed May 15 10:22:52 2024 -0400

    Merge branch 'master' into isa_ext

commit 2502672f19ba8c3bd3888ad8668368d813e629a4
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu May 2 16:40:11 2024 -0400

Implement @jobermayr patch to half precision RISC floating point
support

commit e9b6e414a1809a216dcbb9bb5d10d749452abd4c
Merge: 5f4b333ba4 c014e6851f
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Apr 24 08:55:34 2024 -0400

    Merge branch 'master' into isa_ext

commit 5f4b333ba4f968ff11fc2a74e69075bedb846536
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Apr 23 09:15:26 2024 -0400

Add basic support for RISCV half precision floating point extension
znh

    * conversion semantics are only approximate

commit 61790bff4f2d2212af82191f6fa4ce3d694997b5
Merge: c926250c7d 6c60bd0313
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sun Apr 14 09:45:04 2024 -0400

    Merge branch 'master' into isa_ext

commit c926250c7da996ce140c4bff83c716dfc3036dda
Merge: ce900719f6 8ee5734927
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Apr 11 07:50:26 2024 -0400

    Merge branch 'master' into isa_ext

commit ce900719f62f22624c9cd68b6bd46ac62141046d
Merge: d882bb37fb 0f3351ab6a
Author: thixotropist <97179149+thixotropist@users.noreply.github.com>
Date:   Wed Mar 27 14:52:44 2024 -0400

    Merge branch 'NationalSecurityAgency:master' into isa_ext

commit d882bb37fb5a76ab9b6b1ec14abde017b63196a8
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat Mar 23 11:51:58 2024 -0400

    Add semantics for riscv bit manipulation extension instructions

commit dccd2d119c39734463fd5e1c462de8f05866416f
Merge: 7c52c75cdf 7360b14db0
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Mar 20 08:15:46 2024 -0400

    Merge branch 'master' into isa_ext

commit 7c52c75cdf09e6c5c209d7a50b8e92089f588abe
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Mar 20 08:12:53 2024 -0400

    generate *all* non-reserved vset and vsetvli encoded vtypes.

commit 94061d83566d1bad4ed019f0bb2bdc73eec9a049
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Mar 20 08:12:02 2024 -0400

    Correct rori immediate encoding to support larger shifts

commit 3cd630ad67d2f493ee6252a582891a4545d3c6d1
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Fri Mar 15 08:04:42 2024 -0400

    correct vector crypto semantics to include in/out register usage

commit d33dcc5afa710ae00bd4dc7f46f661e6b1b9c4e0
Merge: 896b74419d b82294b537
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Fri Mar 15 07:21:12 2024 -0400

    Merge branch 'master' into isa_ext

commit 896b74419d2d9e531c69cfcdb60cfb7cc367e533
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Fri Mar 15 07:20:02 2024 -0400

    Add RISCV ISA extension pcode and instruction definitions.

commit 8475eca8b6115f1d1d1211885794ef958ddf9e00
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Feb 29 09:52:24 2024 -0500

    Correct return type for vmv.x.s instruction

commit 4ac1dae8316240186bcccd6c161abfd11aa3956e
Merge: 21948c61a9 c5bad0a88f
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Feb 26 08:50:55 2024 -0500

    Merge branch 'master' into isa_ext

commit 21948c61a992c29cf82270c6102ff91eececa343
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Feb 26 08:49:23 2024 -0500

    Correct multiple missing parameters in pcode semantics

commit ffb5411fb3aec03e24eb7dcc928ad779e43f8be0
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Fri Feb 23 10:39:20 2024 -0500

    Add missing pcode op

commit ba97798c021b44843bcaddac1cd8b18e538612cb
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Feb 21 15:24:07 2024 -0500

    Correct multiple typos in vector pcode semantics

commit f81e17ab88a53b4c90dcb9507b90c40b22bab1b8
Merge: 761dc9ffa6 6d6d0317e2
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Feb 19 10:29:32 2024 -0500

    Merge branch 'master' into isa_ext

commit 761dc9ffa6a8ac8f9177bf5795720d1f130c6bda
Merge: af558cf78b b3ef02212c
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Feb 14 13:32:22 2024 -0500

    Merge branch 'master' into isa_ext

commit af558cf78b57ce1431f7ac7d1d77a87d11923724
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Feb 14 13:30:02 2024 -0500

    remove unused pcodeop

commit 2feeadf8ce8f4aa97487ad702a6aaf6e971fed87
Merge: f266bdef1b 33f90ba1c1
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Feb 5 14:01:08 2024 -0500

    Merge branch 'master' into isa_ext

commit f266bdef1bae56dc55f78dd60ff31426bf035090
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Feb 5 14:00:38 2024 -0500

    add semantics to sh*add and rework vector pcode ops and semantics

commit 20f0331efc75017a35d120b8bfd219d3887e629a
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Fri Jan 26 09:46:15 2024 -0500

    Add RISCV sleigh definition for `fence.tso` instruction

commit 1c6942fd2de5cea6b96f635cf88337167bd849a1
Merge: b6e261dc39 e15e12b248
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Jan 22 13:31:57 2024 -0500

    Merge branch 'master' into isa_ext

commit b6e261dc3993eba2af23c0414444078e96d48914
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Jan 22 07:42:39 2024 -0500

    Correct mangled sleigh definition of bext instruction

commit b07fd57cf06691c608422fd9cfcb24eaec3c38ae
Merge: fa1e0f4b58 3410dc1ad9
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Jan 16 08:40:23 2024 -0500

    Merge branch 'master' into isa_ext

commit fa1e0f4b5871f951a3da843692674496c48ff124
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Jan 16 08:38:41 2024 -0500

    make vector store pcodeop less confusing

commit 705afc67d014683e4f99f0900ef55169866bc779
Merge: 435eaa8edc f7013bdfab
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Dec 26 08:14:45 2023 -0500

    Merge branch 'master' into isa_ext

commit 435eaa8edc1568630eee2f85298065b909545d2a
Merge: 1e240639c7 6242fda158
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Dec 13 08:53:20 2023 -0500

    Merge branch 'master' into isa_ext

commit 1e240639c797e47eaf85d107822a3e9e3d1f592b
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Dec 13 08:52:21 2023 -0500

Update support for bit manipulation and compressed instructions; add
a Readme
    concerning ISA extensions

commit a1bbbfabf7c5d8ccabfee9aa66748e036f9f49ac
Merge: c08750cbcc d4861c46ac
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Nov 29 09:47:24 2023 -0500

    Merge branch 'master' into isa_ext

commit c08750cbccb84d9a539d2b45e619ffb277e8fb06
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sun Nov 19 14:59:35 2023 -0500

    Add placeholder pcodeops to riscv vector instructions

commit b0410be6c7ec7dc0b73fb1b7a77f7730a75c00b7
Merge: 24938f5202 f7ed1b493b
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Nov 14 12:57:30 2023 -0500

    Merge branch 'isa_ext' into semantics

commit f7ed1b493b014de428233b6df55e50a34b8597d4
Merge: 282f300b4a ba30f4b0bb
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Nov 14 12:57:11 2023 -0500

    Merge branch 'master' into isa_ext

commit 282f300b4a57be8542612a3c024cc9d42fee05d9
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Nov 6 12:32:04 2023 -0500

Cherrypick 32 bit corrections and thead semantics provided by
@madushan1000

commit 0b559f5458f9269c2b53bcb2128996556cbcfc3c
Merge: 97838cc6ea 7c9aa58aaf
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Oct 30 06:22:46 2023 -0400

    Merge branch 'master' into isa_ext

commit 97838cc6eaca2fdb87b96e873fba5b46af9a9bb0
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Mon Oct 30 06:21:18 2023 -0400

Collect thead extensions into new language definitions RV64GCV_THEAD
and RV32GCV_THEAD

commit 9837b1da6dc0ae8e686005ec8152faf0282f56b3
Merge: b1b7ea77e9 a24bdf30f7
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Oct 26 09:37:17 2023 -0400

    Merge branch 'master' into isa_ext

commit b1b7ea77e931b0bd7a595103982bc6e9f20137c6
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Oct 26 09:36:06 2023 -0400

Add THead vendor-specific RISCV extensions also implemented in
`objdump`.

commit 033123179601374585fcf45632c261f40f1f1dd1
Merge: e60ba503c4 0820fffbea
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Tue Oct 24 09:11:36 2023 -0400

    Merge branch 'master' into isa_ext

commit e60ba503c49ff830f99838a98fa31644639e01a7
Merge: 3a084aae12 fae64a82c0
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sat Oct 14 11:49:04 2023 -0400

    Merge branch 'master' into isa_ext

commit 3a084aae12197b07a07c9ecd810cb60bd42de1eb
Merge: 6d49b16464 d27902dc09
Author: thixotropist <97179149+thixotropist@users.noreply.github.com>
Date:   Thu Oct 12 10:26:45 2023 -0400

    Merge branch 'NationalSecurityAgency:master' into isa_ext

commit 6d49b16464fc78d45d13a9d1f6de2eea854a17e3
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sun Oct 1 12:09:59 2023 -0400

Add pcode semantics for some of the vector instrinsics used in
example code.

commit 24938f52025c4bbb96d8a6dd90bcf4f9207f4fe4
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Sun Oct 1 12:07:32 2023 -0400

Add pcode semantics for some of the vector instrinsics used in
example code.

commit 902689090a5aa030af6819a63b37882bee087478
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Sep 21 07:25:58 2023 -0400

Fill in some scalar bit manipulation and crypto instruction
extensions

commit 245127060e55de74cd621196c394d35d16551452
Merge: 69490d4200 87d9096b63
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Sep 20 08:30:41 2023 -0400

    Merge branch 'master' into isa_ext

commit 69490d42005243bc81106c34673caca91bfbf03b
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Wed Sep 20 08:29:13 2023 -0400

    Add certifications thanks to @jobermayr

commit 85ba3fcc52e65160c4b4a70ef886cedf568b9df8
Author: Gary Hayward <garyhayward@gmail.com>
Date:   Thu Sep 14 13:46:58 2023 -0400

update RISCV-64 sleigh files to current vector, bitmap, and crypto
extensions

thixotropist commented:
@GhidorahRex - try the newly squashed isa_ext branch now.  I moved the original isa_ext branch to a new branch isa_ext_unsquashed just-in-case.
@jobermayr - thanks for the correction and sample commands.  I used git merge rather than patch merge, mostly for fear of how patch would handle any binary or lfs content.  There isn't any in this repo - I'm just paranoid when forcing a push onto github or gitlab.

GhidorahRex commented:
@thixotropist Yep, the new squashed branch is what I was looking for.

GhidorahRex commented:
mask register issues are prevalent throughout SIMD/vector sets. In the AVX512 I tried to mitigate it as best I could through pcode. But with the RISCV and others it's more challenging because the bits don't correspond one-to-one with the bits in the vector registers - they correspond to elements, which aren't always the same size.

thixotropist committed:
Merge branch 'master' into isa_ext

thixotropist committed:
Merge branch 'master' into isa_ext

LukeSerne commented on Ghidra/Processors/RISCV/data/languages/riscv.rv32b.sinc:
```diff
@@ -1,131 +1,149 @@
 
-:clmulr rd, rs1, rs2 is op0006=0x33 & op1214=0x2 & op2531=0x5 & rd & rs1 & rs2 unimpl
+define pcodeop clz;
+:clz rd, rs1 is op0006=0x13 & op1214=0x1 & op2024=0x0 & op2531=0x30 & rd & rs1 {rd=clz(rs1);}
```
I think you can use the `LZCOUNT` pcode op for this

thixotropist committed:
Merge branch 'master' into isa_ext

thixotropist committed:
Merge branch 'master' into isa_ext

thixotropist committed:
restore missing uimm5 definition for RISC-V vector processing

LukeSerne commented:
> Let's see if the developers have any suggestions here. You make some great points regarding the RISC-V bit manipulation extension instructions that map cleanly into existing pcode ops.

Roughly two months have passed without any developer feedback :disappointed:. @thixotropist, would you be open to replacing some of the custom pcode ops in this PR with existing pcode ops?

thixotropist commented:
@LukeSerne - I'd be happy to make those replacements. How would you go about testing the replacement semantics?  Would the 32 bit and 64 bit semantics likely differ?  Do any of those semantics depend on any Control and Status Register variables?

@GhidorahRex mentioned in Issue #9198 that he wants to spend some time soon reviewing this PR, so any updates from us should make the PR easier to review.

LukeSerne commented:
> How would you go about testing the replacement semantics? Would the 32 bit and 64 bit semantics likely differ? Do any of those semantics depend on any Control and Status Register variables?

I'd just base the implementation on the manual (for example https://docs.riscv.org/reference/isa/unpriv/b-st-ext.html). If the manual turns out not to be accurate, that can always be fixed later. Perhaps by comparing the output of the pcode emulator and qemu or some real hardware that implements these extensions.

> [...] any updates from us should make the PR easier to review.

I think replacing some of the custom pcode ops with existing pcode ops shouldn't increase the complexity of this PR by too much. I think it will even improve this PR because the decompiler output will hopefully be better with most of the custom pcode ops removed.

thixotropist committed:
Merge branch 'master' into isa_ext

thixotropist committed:
RISC-V Vector destination registers are often input registers too

Vector ternary instructions need to mark destination registers as input
registers too.
